File: scripts/run-recommendations.ts
/**
 * Run Recommendations Script
 * 
 * Generates personalized recommendations for all users and optionally sends emails.
 * 
 * Usage:
 *   npm run recommend              # Generate recommendations only
 *   npm run recommend -- --send    # Generate and send emails
 */

import { supabase } from '../src/lib/supabase';
import { TwoTowerRecommender, RecommendedEvent } from '../src/lib/recommender';
import { EmailDeliveryService } from '../src/lib/email-delivery';

interface UserRecommendations {
  userId: string;
  email: string;
  name: string;
  recommendations: RecommendedEvent[];
}

async function runRecommendations() {
  const sendEmails = process.argv.includes('--send');
  
  console.log(' Starting Two-Tower Recommendation Pipeline');
  console.log(`   Mode: ${sendEmails ? 'Generate + Send Emails' : 'Generate Only (dry run)'}`);
  console.log('');

  // Initialize recommender
  const recommender = new TwoTowerRecommender({
    topK: 10,
    recencyWeight: 0.3,
    similarityWeight: 0.7,
    maxDaysOld: 90,
  });

  // Load event index
  console.log(' Loading event embeddings...');
  await recommender.loadEventIndex();

  // Fetch all users
  console.log(' Fetching users...');
  const { data: users, error: usersError } = await supabase
    .from('users')
    .select('id, email, name');

  if (usersError || !users) {
    console.error(' Failed to fetch users:', usersError?.message);
    process.exit(1);
  }

  console.log(`   Found ${users.length} users\n`);

  // Generate recommendations for each user
  const allRecommendations: UserRecommendations[] = [];
  let successCount = 0;
  let errorCount = 0;

  for (const user of users) {
    try {
      process.stdout.write(`Processing ${user.name || user.email}... `);
      
      const recommendations = await recommender.getRecommendations(user.id);
      
      allRecommendations.push({
        userId: user.id,
        email: user.email,
        name: user.name || 'User',
        recommendations,
      });

      // Log recommendations to database
      for (let i = 0; i < recommendations.length; i++) {
        const rec = recommendations[i];
        await supabase.from('recommendation_logs').insert({
          user_id: user.id,
          event_id: rec.event_id,
          similarity_score: rec.similarity_score,
          recency_score: rec.recency_score,
          final_score: rec.final_score,
          position: i + 1,
          model_version: 'v2-two-tower',
        });
      }

      successCount++;
      console.log(`✓ ${recommendations.length} recommendations`);

    } catch (error: any) {
      errorCount++;
      console.log(`✗ Error: ${error.message}`);
    }
  }

  // Print summary
  console.log('\n' + '='.repeat(60));
  console.log(' Recommendation Summary');
  console.log('='.repeat(60));
  console.log(`   Users processed: ${successCount}/${users.length}`);
  console.log(`   Errors: ${errorCount}`);
  
  // Statistics
  const allRecs = allRecommendations.flatMap(u => u.recommendations);
  if (allRecs.length > 0) {
    const avgSimilarity = allRecs.reduce((s, r) => s + r.similarity_score, 0) / allRecs.length;
    const avgFinalScore = allRecs.reduce((s, r) => s + r.final_score, 0) / allRecs.length;
    
    console.log(`   Total recommendations: ${allRecs.length}`);
    console.log(`   Avg similarity score: ${avgSimilarity.toFixed(3)}`);
    console.log(`   Avg final score: ${avgFinalScore.toFixed(3)}`);
    
    // Category distribution
    const categoryDist: Record<string, number> = {};
    for (const rec of allRecs) {
      const cat = rec.category || 'unknown';
      categoryDist[cat] = (categoryDist[cat] || 0) + 1;
    }
    
    console.log('\n Category Distribution in Recommendations:');
    for (const [cat, count] of Object.entries(categoryDist).sort((a, b) => b[1] - a[1])) {
      const pct = (count / allRecs.length * 100).toFixed(1);
      console.log(`   ${cat.padEnd(20)} ${count.toString().padStart(4)} (${pct}%)`);
    }
  }

  // Send emails if requested
  if (sendEmails) {
    console.log('\n Sending personalized emails...');
    
    const emailService = new EmailDeliveryService();
    
    // Convert to format expected by email service
    const emailMap = new Map();
    for (const userRecs of allRecommendations) {
      if (userRecs.recommendations.length > 0) {
        // Convert RecommendedEvent to RankedEvent format
        const rankedEvents = userRecs.recommendations.map(rec => ({
          id: rec.event_id,
          title: rec.title,
          description: rec.caption,
          event_date: rec.timestamp,
          location: null,
          category: rec.category,
          tags: null,
          source_url: rec.permalink,
          created_at: rec.timestamp,
          updated_at: rec.timestamp,
          score: rec.final_score * 100, // Scale to match old scoring
          cluster_match: rec.similarity_score,
          urgency_score: rec.recency_score * 30,
        }));
        
        emailMap.set(userRecs.userId, rankedEvents);
      }
    }

    await emailService.sendNewsletters(emailMap);
    console.log(' Email sending complete!');
  }

  // Print sample recommendations
  console.log('\n' + '='.repeat(60));
  console.log(' Sample Recommendations (first 3 users)');
  console.log('='.repeat(60));
  
  for (const userRecs of allRecommendations.slice(0, 3)) {
    console.log(`\n👤 ${userRecs.name} (${userRecs.email})`);
    
    for (let i = 0; i < Math.min(3, userRecs.recommendations.length); i++) {
      const rec = userRecs.recommendations[i];
      console.log(`   ${i + 1}. ${rec.title.substring(0, 50)}...`);
      console.log(`      Score: ${rec.final_score.toFixed(3)} | Category: ${rec.category || 'unknown'}`);
      console.log(`      Reason: ${rec.reason}`);
    }
  }

  console.log('\n Recommendation pipeline complete!');
}

runRecommendations().catch(console.error);

File: python/scripts/run_recommendations.py
"""
Run Recommendations Script

Generates personalized recommendations for all users.
Run: python -m python.scripts.run_recommendations
"""
import argparse
from dotenv import load_dotenv
import sys
import os

# Add parent directory to path for imports
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from lib.supabase_client import supabase
from lib.recommender import recommender
from lib.email_delivery import email_delivery_service

# Load environment variables
load_dotenv()


def main():
    parser = argparse.ArgumentParser(description="Generate recommendations")
    parser.add_argument("--send", action="store_true", help="Send emails after generating")
    parser.add_argument("--limit", type=int, default=10, help="Number of users to process")
    args = parser.parse_args()
    
    print("🎯 Starting Two-Tower Recommendation Pipeline")
    print(f"   Mode: {'Generate and Send' if args.send else 'Generate Only (dry run)'}")
    print()
    
    # Load event embeddings
    print("📚 Loading event embeddings...")
    recommender.load_event_index()
    print(f"Loaded {recommender.event_index.size()} events into vector index")
    
    # Fetch users
    print("👤 Fetching users...")
    result = supabase.table("users").select("id, email, name").limit(args.limit).execute()
    users = result.data or []
    print(f"   Found {len(users)} users")
    print()
    
    # Generate recommendations for each user
    recommendations_by_user = {}
    
    for user in users:
        try:
            recs = recommender.get_recommendations(user["id"])
            recommendations_by_user[user["id"]] = [
                {
                    "id": r.event_id,
                    "title": r.title,
                    "category": r.category,
                    "score": r.final_score
                }
                for r in recs
            ]
            print(f"Processing {user.get('name', user['email'])}... ✓ {len(recs)} recommendations")
        except Exception as e:
            print(f"Processing {user.get('name', user['email'])}... ✗ Error: {e}")
            recommendations_by_user[user["id"]] = []
    
    print()
    print(f"📊 Generated recommendations for {len(users)} users")
    
    # Send emails if requested
    if args.send:
        print()
        print("📧 Sending emails...")
        result = email_delivery_service.send_newsletters(recommendations_by_user)
        print(f"   Sent: {result['success']}, Failed: {result['failure']}")
    
    print()
    print("✅ Recommendation pipeline complete!")


if __name__ == "__main__":
    main()
